Locked jobs

ggueret opened the pull request:
- Added 'expired' status for the jobs who were unable to lock
- Added 'locked_job' bool option to `mrq.task.Task`, to lock a job execution by his task path.
- Mark `mrq.basetasks.cleaning.[RequeueInterruptedJobs, RequeueRetryJobs, RequeueStartedJobs, RequeueRedisStartedJobs, RequeueLostJobs, MigrateKnownQueues, CleanKnownQueues, ]`  as locked jobs.

sylvinus commented:
- `locked_job` is probably not the best name
 - `lock_timeout` is not explicit enough (this lock should have a specific name to avoid confusion)
 - The locking code is not safe enough
 - I'm not (yet) convinced we need to add a new status `expired` just for this feature.

sylvinus commented on mrq/basetasks/cleaning.py:
```diff
@@ -10,6 +10,8 @@ class RequeueInterruptedJobs(Task):
 
     """ Requeue jobs that were marked as status=interrupt when a worker got a SIGTERM. """
 
+    max_concurrency = 1
```
"maxconcurrency" serait plus facile à grepper ?

ggueret replied:
c'est par respect de la sémantique existante, avec par exemple le `Job.max_retries`

sylvinus commented on mrq/config.py:
```diff
@@ -391,6 +391,13 @@ def add_parser_args(parser, config_type):
                  'to dequeue them in command-line order.')
 
+        parser.add_argument(
+            '--lock_timeout',
```
--maxconcurrency_timeout

sylvinus commented on mrq/config.py:
```diff
@@ -391,6 +391,13 @@ def add_parser_args(parser, config_type):
+            default=1.,
+            type=float,
+            action='store',
+            help='Max seconds to wait for a lock before changing the job status to expired.')
```
text to update

sylvinus commented on mrq/job.py:
```diff
@@ -274,7 +279,22 @@ def perform(self):
+                pipeline = context.connections.redis.pipeline()
+                pipeline.incr(self.redis_concurrency_key)
+                pipeline.expireat(self.redis_concurrency_key, int(time.time()) + self.timeout)
+                current = pipeline.execute()[0]
```
this is not safe enough - with maxconcurrency=2, if a worker dies here, the count will stay at a minimum at 1, and if there are still tasks happening, the key will never expire.

ggueret commented:
- Added 'maxconcurrency' status for the jobs who reach the max_concurrency limit.
- Replaced 'locked_job' by 'max_concurrency' who now take an integer value.
- Mark mrq.basetasks.cleaning.[RequeueInterruptedJobs, RequeueRetryJobs, RequeueStartedJobs, RequeueRedisStartedJobs, RequeueLostJobs, MigrateKnownQueues, CleanKnownQueues] with a concurrency of 1 per `Job.timeout`.
